python/cuspatial/cuspatial/io/geoseries_reader.py

```diff
@@ -0,0 +1,284 @@
+                current = offsets["polygons"]["polygons"][-1]
+                offsets["polygons"]["polygons"].append(num_rings + current)
+            elif isinstance(geometry, MultiPolygon):
+                current = offsets["polygons"]["polygons"][-1]
+                offsets["polygons"]["mpolys"].append(
+                    len(offsets["polygons"]["polygons"]) - 1
+                )
+                for poly in geometry:
+                    current = offsets["polygons"]["polygons"][-1]
+                    num_rings = 1
+                    rings_current = offsets["polygons"]["rings"][-1]
+                    offsets["polygons"]["rings"].append(
+                        len(poly.exterior.coords) * 2 + rings_current
+                    )
+                    for interior in poly.interiors:
+                        rings_current = offsets["polygons"]["rings"][-1]
+                        offsets["polygons"]["rings"].append(
+                            len(interior.coords) * 2 + rings_current
+                        )
+                        num_rings = num_rings + 1
+                    offsets["polygons"]["polygons"].append(num_rings + current)
+                offsets["polygons"]["mpolys"].append(
+                    len(offsets["polygons"]["polygons"]) - 1
+                )
+        return offsets
+
+    def _read_geometries(self, geoseries, offsets):
+        """
+        Creates a set of buffers sized to fit all of the geometries and
+        iteratively populates them with geometry coordinate values.
+
+        Parameters
+        ----------
+        geoseries : A GeoPandas GeoSeries object.
+        offsets : The set of offsets that correspond to the geoseries argument.
+        """
+        buffers = {
+            "points": cudf.Series(cp.zeros(offsets["points"][-1])),
+            "multipoints": cudf.Series(cp.zeros(offsets["multipoints"][-1])),
+            "lines": cudf.Series(cp.zeros(offsets["lines"][-1])),
+            "polygons": {
+                "polygons": cudf.Series(
+                    cp.zeros(len(offsets["polygons"]["polygons"]))
+                ),
+                "rings": cudf.Series(
+                    cp.zeros(len(offsets["polygons"]["rings"]))
+                ),
+                "coords": cudf.Series(
+                    cp.zeros(offsets["polygons"]["rings"][-1])
+                ),
+            },
+        }
+        read_count = {
+            "points": 0,
+            "multipoints": 0,
+            "lines": 0,
+            "polygons": 0,
+        }
+        inputs = []
+        input_types = []
+        input_lengths = []
+        for geometry in geoseries:
+            if isinstance(geometry, Point):
+                # write a point to the points buffer
+                # increase read_count of points pass
+                p = cp.array(geometry)
+                i = read_count["points"] * 2
+                buffers["points"][i] = p[0]
+                buffers["points"][i + 1] = p[1]
```

Comment:
Is constructing a cupy array from a `Point` object efficient?

What is the expected types of `buffers["points"][i]` and `buffers["points"][i + 1]`?

Should we really be using cupy arrays or arrays at all here? Should we just be directly interfacing with the `Point` object?

Reply:
I'm not sure. The two points in `geometry` are going to be copied from host to device in this step. In this particular case I think it could be efficient to access the `Point` values directly, but not a whole lot more efficient, most likely. I am going to look into it and update it to the most efficient expression I can come up with.

Reply:
It looks like Shapely objects are all `float64`, so all these buffers are `float64`.

Reply:
This is in a big loop though, so all of those H2D copies will seriously add up and slow things down. We should collecting things into nicely organized buffers on the CPU side and then doing bulk copies to the GPU at the end.

Reply:
Good point. I guess I should be doing that with all of the copies - put into RAM buffers first, then copy those whole buffers to GPU.